services-marionette/firefox-services-tests/apps/base.py

```diff
@@ -0,0 +1,88 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+from marionette_driver import expected, Wait
+from marionette_driver.errors import NoSuchElementException
+
+import re
+
+
+class Base(object):
+
+    def __init__(self, marionette):
+        self.marionette = marionette
+        self.CHROME = 'chrome'
+        self.CONTENT = 'content'
+        self.set_context(self.CONTENT)
+
+    def launch(self, url):
+        if url is not None:
+            regex = re.compile(
+                r'^(?:http|ftp)s?://'
+                r'(?:(?:[A-Z0-9](?:[A-Z0-9-]{0,61}[A-Z0-9])?\.)'
+                r'+(?:[A-Z]{2,6}\.?|[A-Z0-9-]{2,}\.?)|'
+                r'localhost|'
+                r'\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3})'
+                r'(?::\d+)?'
+                r'(?:/?|[/?]\S+)$', re.IGNORECASE)
```

Comment:
LOL! I would love to say that I am the originator of such a crazy regex. I augmented my original regex with some hints I found online.